### PR TITLE
"Unexpected type: " error #197 fix

### PR DESCRIPTION
**Pull request by clangelov**

### CODE REVIEW

**Comment by Miragecoder** [comment]
Linked to #197.

**Comment by Miragecoder** [comment]
Hi-ya, @tlil!

We've been running our GraphQL.NET/Conventions implementation - consisting on like 30 or so queries/mutations/subscriptions - on a custom build based on the changes proposed in this PR. In addition to our 'node'-queries running successfully without the `Unexpected Type: `-error we haven't seen any negative impact so far.

We do wonder whether the filtering applied in `GraphTypeAdapter.OmitNonRootTypes` here is conceptually sound. What do you think? 😃

File: src/GraphQL.Conventions/Adapters/GraphTypeAdapter.cs [hunk 1]
```diff
@@ -29,25 +29,45 @@ public ISchema DeriveSchema(GraphSchemaInfo schemaInfo)
             var interfaces = types
                 .Where(t => t.IsInterfaceType && !t.IsInputType && t.PossibleTypes.Any())
                 .GroupBy(t => t.Name)
-                .Select(g => g.First());
-            var possibleTypes = interfaces
+                .Select(g => g.First());
```
Comment (tlil) [inline src/GraphQL.Conventions/Adapters/GraphTypeAdapter.cs]: Nitpick - extra EOL spacing :)

File: src/GraphQL.Conventions/Adapters/GraphTypeAdapter.cs [hunk 2]
```diff
@@ -29,25 +29,45 @@ public ISchema DeriveSchema(GraphSchemaInfo schemaInfo)
...
-            var possibleTypes = interfaces
+                .Select(g => g.First());                
+            var myPossibleTypes = interfaces
                 .Where(t => !t.IsIgnored)
                 .SelectMany(t => t.PossibleTypes)
-                .GroupBy(t => t.Name)
-                .Select(g => g.First());
+                .Select(x => OmitNonRootTypes(x))
+                .Where(x => x != null)
+                .ToArray();
```
Comment (tlil) [inline src/GraphQL.Conventions/Adapters/GraphTypeAdapter.cs:43]: By not grouping, isn't there a chance that you get duplicate references here if the same type appear as "possible type" for more than one interface?

File: src/GraphQL.Conventions/Adapters/GraphTypeAdapter.cs [hunk 3]
```diff
@@ -306,21 +327,21 @@ private IGraphType ConstructOutputType(GraphTypeInfo typeInfo)
...
         }
 
-        private static object UnwrapObject(object obj)
+        private object UnwrapObject(object obj)
         {
-            if (obj is INonNull @null)
+            if (obj is INonNull)
             {
-                obj = @null.ObjectValue;
+                obj = ((INonNull)obj)?.ObjectValue;
```
Comment (tlil) [inline src/GraphQL.Conventions/Adapters/GraphTypeAdapter.cs]: Why change these pattern matches back?

File: src/GraphQL.Conventions/Adapters/GraphTypeAdapter.cs [hunk 4]
```diff
@@ -217,7 +238,7 @@ private IGraphType CacheType(GraphTypeInfo typeInfo, IGraphType graphType)
         private TType CreateTypeInstance<TType>(Type type)
         {
             var obj = Activator.CreateInstance(type);
-            return obj is TType castType ? castType : default;
+            return obj is TType ? (TType)obj : default;
```
Comment (tlil) [inline src/GraphQL.Conventions/Adapters/GraphTypeAdapter.cs]: Why?

File: src/GraphQL.Conventions/GraphQL.Conventions.csproj [hunk 5]
```diff
@@ -2,7 +2,7 @@
 
   <PropertyGroup>
     <Description>GraphQL Conventions for .NET</Description>
-    <VersionPrefix>2.5.3</VersionPrefix>
+    <VersionPrefix>2.5.4</VersionPrefix>
```
Comment (tlil) [inline src/GraphQL.Conventions/GraphQL.Conventions.csproj:5]: You'd need to update `CommonAssemblyInfo.cs` as well.

File: src/GraphQL.Conventions/Adapters/GraphTypeAdapter.cs [hunk 6]
```diff
@@ -29,25 +29,45 @@ public ISchema DeriveSchema(GraphSchemaInfo schemaInfo)
...
                 Query = DeriveOperationType(schemaInfo.Query),
                 Mutation = DeriveOperationType(schemaInfo.Mutation),
                 Subscription = DeriveOperationType(schemaInfo.Subscription),
             };
-            schema.RegisterTypes(possibleTypes
-                .Where(t => !t.IsIgnored && !t.Interfaces.Any(i => i.IsIgnored))
-                .Select(t => DeriveType(t).GetType())
-                .ToArray());
+            // Instead of using Type[] we are now using IGraphType[] for the schema registration
+            schema.RegisterTypes(myPossibleTypes);
```
Comment (tlil) [inline src/GraphQL.Conventions/Adapters/GraphTypeAdapter.cs]: By omitting `!t.Interfaces.Any(i => i.IsIgnored)`, does that mean that you're no longer checking the underlying interfaces, and whether any of those are marked as ignored?

**Review by tlil — COMMENTED** [review]
Thanks for this - just a few comments :-)

Since you mention #197, could you add the explicit case in that issue as a test as well, please?

**Comment by Miragecoder** [comment]
Thanks for your remarks, @tlil - We'll look into addressing them asap!

**Comment by Miragecoder** [comment]
Hey @tlil - Thanks for your feedback! I believe your comments are addressed now; we're wondering about how you feel about the changes. Would this be sufficient for a merge? 😃

**Review by tlil — APPROVED** [review]
Yep, thanks! :)

**Pull request merged**